Orm/Xtensive.Orm/Orm/Rse/Transformation/ColumnMappingInspector.cs

```diff
@@ -31,57 +37,72 @@ public virtual CompilableProvider RemoveRedundantColumns()
 
     protected override Provider VisitInclude(IncludeProvider provider)
     {
-      int sourceLength = provider.Source.Header.Length;
+      var sourceLength = provider.Source.Header.Length;
       mappings[provider.Source] = Merge(mappings[provider].Where(i => i < sourceLength), provider.FilteredColumns);
       var source = VisitCompilable(provider.Source);
       mappings[provider] = Merge(mappings[provider], mappings[provider.Source]);
-      if (source==provider.Source)
+      if (source == provider.Source) {
         return provider;
-      var filteredColumns = provider.FilteredColumns.Select(el => mappings[provider].IndexOf(el)).ToArray();
+      }
+      var filteredColumns = provider.FilteredColumns
+        .Select(el => mappings[provider].IndexOf(el))
+        .ToArray(provider.FilteredColumns.Count);
       return new IncludeProvider(source, provider.Algorithm, provider.IsInlined,
         provider.FilterDataSource, provider.ResultColumnName, filteredColumns);
     }
 
     protected override Provider VisitSelect(SelectProvider provider)
     {
       var requiredColumns = mappings[provider];
-      var remappedColumns = requiredColumns.Select(c => provider.ColumnIndexes[c]).ToList();
+      var remappedColumns = requiredColumns
+        .Select(c => provider.ColumnIndexes[c])
+        .ToList(requiredColumns.Count);
+
       mappings[provider.Source] = remappedColumns;
       var source = VisitCompilable(provider.Source);
       var sourceMap = mappings[provider.Source];
-      var columns = provider.ColumnIndexes
-        .Select((i, j) => new Pair<int>(sourceMap.IndexOf(i), j))
-        .Where(i => i.First >= 0)
-        .ToList();
-      mappings[provider] = columns.Select(c => c.Second).ToList();
-      var indexColumns = columns.Select(c => c.First).ToList();
-      if (source==provider.Source)
-        return provider;
-      return new SelectProvider(source, indexColumns.ToArray());
+
+      var indexColumns = new List<int>();
+      var newMappings = new List<int>();
```

Comment:
Shouldn't we specify initial capacity here? It might be equal to the `provider.ColumnIndexes.Count`.

Reply:
Done